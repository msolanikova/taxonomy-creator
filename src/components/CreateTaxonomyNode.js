import React, {useState} from 'react';
import {InputLabel, MenuItem, Select} from "@mui/material";
import Button from "@mui/material/Button";

const CreateTaxonomyNode = ({valuesByType, addNodeToTaxonomy}) => {
    const [taxoNode, setTaxoNode] = useState({});

    const renderItemsForType = (nodes, type) => {
        const nodesCopy = [{name: ""}, ...nodes];
        const items = nodesCopy.map(node => (
            // <MenuItem key={node.name} value={node.name}>{node.name}</MenuItem>
            <option key={node.name ? node.name : `${type}-null`} value={node.name}>{node.name}</option>
        ));
        return items;
    }

    const handleSelect = (type, value) => {
        const taxoNodeCopy = {...taxoNode};
        if(value === "") {
            delete taxoNodeCopy[type];
        } else {
            taxoNodeCopy[type] = value;
        }
        setTaxoNode(taxoNodeCopy);
    }

    const renderSelectsForTypes = (valuesByType) => {
        return Object.keys(valuesByType).map(type => {
            const nodes = valuesByType[type];
            const items = renderItemsForType(nodes, type);
            return (
                <div key={type}>
                    <InputLabel id="demo-simple-select-label">{type}</InputLabel>
                    <Select
                        labelId="demo-simple-select-label"
                        id={type}
                        native={true}
                        value={taxoNode[type] ?? ""}
                        label={type}
                        onChange={(ev) => handleSelect(type, ev.target.value)}
                    >
                        {items}
                    </Select>
                </div>
            );
        });
    }

    const addTaxoNode = () => {
        addNodeToTaxonomy(taxoNode);
        setTaxoNode({})
    }

    return (
        <div>
            <h2>Create Taxonomy Node</h2>
            {renderSelectsForTypes(valuesByType)}
            <Button variant="contained" disabled={Object.keys(taxoNode).length === 0} onClick={addTaxoNode}>Add Taxonomy Node</Button>
        </div>
    );
};

export default CreateTaxonomyNode;
