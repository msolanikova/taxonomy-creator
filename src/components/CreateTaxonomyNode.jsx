import React, {useState} from 'react';
import {InputLabel, Select} from "@mui/material";
import Button from "@mui/material/Button";
import {connect} from "react-redux";
import taxonomyService from "../services/taxonomy.service";

const CreateTaxonomyNode = ({types, valuesByType, taxonomy, dispatch}) => {
    const [taxoNode, setTaxoNode] = useState({});

    const addNodeToTaxonomy = (node) => {
        const taxoCopy = taxonomyService.addNodeToTaxonomy(node, taxonomy, types);
        dispatch({
            type: "CHANGE_TAXONOMY", data: {
                taxonomy: taxoCopy
            }
        });
    }

    const renderItemsForType = (values, type) => {
        const valuesCopy = ["", ...values];
        const items = valuesCopy.map(value => (
            <option key={value ? value : `${type}-null`} value={value}>{value}</option>
        ));
        return items;
    }

    const handleSelect = (type, value) => {
        const taxoNodeCopy = {...taxoNode};
        if (value === "") {
            delete taxoNodeCopy[type];
        } else {
            taxoNodeCopy[type] = value;
        }
        setTaxoNode(taxoNodeCopy);
    }

    const renderSelectsForTypes = (valuesByType) => {
        return Object.keys(valuesByType).map(type => {
            const values = valuesByType[type];
            const items = renderItemsForType(values, type);
            return (
                <div key={type}>
                    <InputLabel id="demo-simple-select-label">{type}</InputLabel>
                    <Select
                        labelId="demo-simple-select-label"
                        id={type}
                        native={true}
                        value={taxoNode[type] ?? ""}
                        label={type}
                        onChange={(ev) => handleSelect(type, ev.target.value)}
                    >
                        {items}
                    </Select>
                </div>
            );
        });
    }

    const addTaxoNode = () => {
        addNodeToTaxonomy(taxoNode);
        setTaxoNode({})
    }

    return (
        <div>
            <h2>Create Taxonomy Node</h2>
            {renderSelectsForTypes(valuesByType)}
            <Button variant="contained" disabled={Object.keys(taxoNode).length === 0} onClick={addTaxoNode}>Add Taxonomy Node</Button>
        </div>
    );
};

// export default CreateTaxonomyNode;
export default connect((state, props) => {
    return {
        types: state.types.types,
        valuesByType: state.valuesByType,
        taxonomy: state.taxonomy
    }
})(CreateTaxonomyNode);
