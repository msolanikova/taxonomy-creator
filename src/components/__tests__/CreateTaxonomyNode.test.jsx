/**
 * @jest-environment jsdom
 */

import {fireEvent, render, screen, within} from "@testing-library/react";
import CreateTaxonomyNode from "../CreateTaxonomyNode";
import React from 'react';
import {Provider} from "react-redux";
import configureMockStore from "redux-mock-store";
import taxonomyServiceMock from "../../services/taxonomy.service";

const mockStore = configureMockStore();
taxonomyServiceMock.addNodeToTaxonomy = jest.fn();

describe("CreateTaxonomyNode component", () => {
    it("should render selects for each type correctly", () => {
        const state = {
            types: ["type1", "type2"],
            valuesByType: {"type1": ["value1","value2"], "type2": ["value3"]},
            taxonomy: {type: "root", path: "/root", name: "Root", children: []}
        }

        const store = mockStore(state);
        const view = render(<Provider store={store}><CreateTaxonomyNode/></Provider>);

        const allSelects = screen.getAllByRole("combobox");
        expect(allSelects).toHaveLength(2);
        const type1Select = allSelects[0];
        const type2Select = allSelects[1];
        expect(screen.getAllByRole("button")).toHaveLength(1);

        const type1Options = within(type1Select).getAllByRole('option');
        const type2Options = within(type2Select).getAllByRole('option');
        expect(type1Options).toHaveLength(3);
        expect(type2Options).toHaveLength(2);
        expect(view).toMatchSnapshot();
    });

    it("should render no selects if there is no type provided", () => {
        const state = {
            types: [],
            valuesByType: {},
            taxonomy: {type: "root", path: "/root", name: "Root", children: []}
        }

        const store = mockStore(state);
        const view = render(<Provider store={store}><CreateTaxonomyNode/></Provider>);

        expect(screen.queryAllByRole("combobox")).toHaveLength(0);
        expect(screen.queryAllByRole("button")).toHaveLength(1);
        expect(view).toMatchSnapshot();
    });

    it("should add new node to taxonomy on hitting 'Create Taxonomy Node' button", async () => {
        const state = {
            types: ["type1", "type2"],
            valuesByType: {"type1": ["value1","value2"], "type2": ["value3"]},
            taxonomy: {type: "root", path: "/root", name: "Root", children: []}
        }

        const store = mockStore(state);
        render(<Provider store={store}><CreateTaxonomyNode/></Provider>);

        fireEvent.change(screen.queryAllByRole("combobox")[0], { target: { value: "value2" } })
        fireEvent.change(screen.queryAllByRole("combobox")[1], { target: { value: "value3" } })
        fireEvent.click(screen.getByRole("button"));

        expect(store.getActions()).toHaveLength(1);
        expect(store.getActions()[0].type).toBeDefined();
        expect(store.getActions()[0].type).toBe("CHANGE_TAXONOMY");
        expect(taxonomyServiceMock.addNodeToTaxonomy).toHaveBeenCalledTimes(1);
        expect(taxonomyServiceMock.addNodeToTaxonomy.mock.calls[0][0]).toEqual({"type1": "value2", "type2": "value3"});
    });
});
