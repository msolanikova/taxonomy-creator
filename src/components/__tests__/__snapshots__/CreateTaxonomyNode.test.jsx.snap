// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTaxonomyNode component should render no selects if there is no type provided 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h2>
          Create Taxonomy Node
        </h2>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Add Taxonomy Node
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2>
        Create Taxonomy Node
      </h2>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        Add Taxonomy Node
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CreateTaxonomyNode component should render selects for each type correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h2>
          Create Taxonomy Node
        </h2>
        <div>
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
            id="demo-simple-select-label"
          >
            type1
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <select
              class="MuiNativeSelect-select MuiNativeSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-ffg8md-MuiNativeSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="type1"
            >
              <option
                value=""
              />
              <option
                value="value1"
              >
                value1
              </option>
              <option
                value="value2"
              >
                value2
              </option>
            </select>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiNativeSelect-icon MuiNativeSelect-iconOutlined css-1g12qau-MuiSvgIcon-root-MuiNativeSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  type1
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div>
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
            id="demo-simple-select-label"
          >
            type2
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <select
              class="MuiNativeSelect-select MuiNativeSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-ffg8md-MuiNativeSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="type2"
            >
              <option
                value=""
              />
              <option
                value="value3"
              >
                value3
              </option>
            </select>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiNativeSelect-icon MuiNativeSelect-iconOutlined css-1g12qau-MuiSvgIcon-root-MuiNativeSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  type2
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Add Taxonomy Node
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2>
        Create Taxonomy Node
      </h2>
      <div>
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
          id="demo-simple-select-label"
        >
          type1
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <select
            class="MuiNativeSelect-select MuiNativeSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-ffg8md-MuiNativeSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="type1"
          >
            <option
              value=""
            />
            <option
              value="value1"
            >
              value1
            </option>
            <option
              value="value2"
            >
              value2
            </option>
          </select>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiNativeSelect-icon MuiNativeSelect-iconOutlined css-1g12qau-MuiSvgIcon-root-MuiNativeSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                type1
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div>
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
          id="demo-simple-select-label"
        >
          type2
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <select
            class="MuiNativeSelect-select MuiNativeSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-ffg8md-MuiNativeSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="type2"
          >
            <option
              value=""
            />
            <option
              value="value3"
            >
              value3
            </option>
          </select>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiNativeSelect-icon MuiNativeSelect-iconOutlined css-1g12qau-MuiSvgIcon-root-MuiNativeSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                type2
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        Add Taxonomy Node
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
